### PR TITLE
Refresh Token Support & Auth Restructure

### LINKED ISSUE #3: Refresh Token Support
We should support the use of refresh tokens so that users don't have to interact with browser login every hour or so

**Issue originally raised by Aidan Kovacic.**

### PR DESCRIPTION
**Pull request by mattphelps-8451**
This PR restructures Auth to make it more extensible, so that other auth services can be added more easily later.
It also adds support for storing and using refresh and access tokens with Azure.
When the Access Token expires, the Refresh Token will be exchanged for a new one, along with a new refresh token and automatically stored.

This allows for a seamless experience, where the user only has to hopefully use the browser login once, or if the Refresh Token expires. Though, if the refresh token does fail for whatever reason, it will ask the user to do browser auth.

This has now been tested. To test it yourself, set scope to "api://SERVICE_NAME/read".
In SERVICE_NAME, you'll need to add "api://SERVICE_NAME" as an extra audience, and set the App Reg Client ID to the one for app-SERVICE_NAMEDev. There probably should be a change in SERVICE_NAME to make it so that it's not hardcoded.

To test refresh token exchanging, on MacOS, go to Users//Library/Preferences/rs.flight-cli-auth/flight-cli-auth.toml and set the expires_on to a time in the past.

**PR and commits originally raised by Aidan**. This would resolve #3.

Note: This was rebased by Matt Phelps with the following: `git rebase --onto ossmain-refresh-token originmain originmain-refresh-token` where ossmain-refresh-token is current branch, originmain is internal 84.51 main branch, originmain-refresh-token is internal 84.51 feature branch.

### CODE REVIEW

File: src/auth/azure.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        client_id,
+        None,
+        &tenant_id,
+        Url::parse("http://localhost:47471")?,
+        &*scope,
+    );
+    println!("\nOpen this URL in a browser:\n{}", c.authorize_url);
+
+    // Using a naive, blocking redirect server is fine for our case, as it should block anyway
+    let code = development::naive_redirect_server(&c, 47471)?;
```
Comment (aidankovacic-8451) [inline src/auth/azure.rs:44]: I originally felt that this was okay for our purposes, but the more I think about it, I feel we should look for a better solution. There have been some updates to azure-identity, so perhaps we could take another look and see if we can make it work here. Perhaps Client Credentials Flow?

File: Cargo.toml [hunk 2]
```diff
@@ -30,18 +30,18 @@ edition = "2021"
...
 
 [dependencies]
-arrow = "19.0.0"
-arrow-flight = "19.0.0"
-azure_core = "0.3.0"
-azure_identity = "0.4.0"
+arrow = "20.0.0"
+arrow-flight = "20.0.0"
+azure_core = "0.4.0"
+azure_identity = "0.5.0"
```
Comment (aidankovacic-8451) [inline Cargo.toml:36]: These have had some updates since this was written. Through testing, I've noticed some breaking changes though, so it may be best left for another PR. This may also be useful, as mentioned in another comment.

File: src/auth.rs [hunk 3]
```diff
@@ -153,6 +161,7 @@ impl Default for AuthStore {
             token: AccessToken::new("".to_string()),
             expiry_time: DateTime::<Utc>::MIN_UTC,
             identity_provider: IdentityProvider::None,
+            refresh_token: RefreshToken::new("".to_string()),
```
Comment (aidankovacic-8451) [inline src/auth.rs:164]: We may want to store refresh tokens more securely in the future. This was a quick-and-dirty way of doing it at the time, but I'm not sure saving the tokens in plain text on the hard drive is following good security practices...

**Review by aidankovacic-8451 — COMMENTED** [review]
Overall, this PR is a definite improvement over what is currently in main.

As mentioned in the review, I do think we should find an alternative to using the development redirect server. I believe that can be done along with updating the azure crates to their latest versions.

We might also want to begin thinking about documenting certain things, such as where configs are located and how to use the CLI with different auth providers.

For now though, I believe this to be fine. Though, I would like to hear your thoughts on the feedback I have given above.

File: src/auth/azure.rs (hunk 1 above)
Comment (mattphelps-8451) [reply src/auth/azure.rs:44]: Hi @KovacicAi, can you provide a link to those updates? I assumed we'd want to stick with the authorization code grant type. https://oauth2.thephpleague.com/authorization-server/which-grant/

Have you seen the Azure Cli Credential? That is one option where we require CLI users to retrieve tokens via the Az CLI. https://docs.rs/azure_identity/latest/azure_identity/struct.AzureCliCredential.html Perhaps that could simplify the auth logic a good deal.

File: Cargo.toml (hunk 2 above)
Comment (mattphelps-8451) [reply Cargo.toml:36]: Which do you recommend as updates and which should we keep as is? Sounds like you want to bump azure_identity at least to latest 0.6.0 (from your other comments)

File: src/auth.rs (hunk 3 above)
Comment (mattphelps-8451) [reply src/auth.rs:164]: Made an issue here for later https://github.com/8451/arrow-flight-cli/issues/7

**Comment by mattphelps-8451** [comment]
@KovacicAi Thanks for the responses... I really appreciate your continued investment in this effort!

File: Cargo.toml (hunk 2 above)
Comment (aidankovacic-8451) [reply Cargo.toml:36]: I believe we should work to update azure_identity to 0.6.0. 
The breaking change should be relatively easy to fix. 0.5.0 changed exchange() to now require an HTTP client. It also changes its oauth2 crate dependency.
The changelogs for azure_identity (along with the other Azure SDK for Rust crates) can be found [here](https://github.com/Azure/azure-sdk-for-rust/releases).
There are other updates to other crates as well, such as Arrow Flight now being at 25.0.0, though I haven't really looked into it, I feel that we should do our best to keep up to date. This could all be done in another PR imo.

File: src/auth/azure.rs (hunk 1 above)
Comment (aidankovacic-8451) [reply src/auth/azure.rs:44]: Hi @mattphelps-8451 ! Sure thing.
I was thinking along the lines of the Azure CLI Credential. 
I wasn't able to get that to work at the time, but I feel that we should at least provide the option of using it as it could make the use of the CLI even easier for people who are logged into the Azure CLI.
My main idea is that we default to the CLI credential, and use Auth Code flow as a fallback, and we could possibly add a flag to force either one. I think that could be made an issue though.

I know I'm flip-flopping, but I think this auth code flow is fine as-is right now, though it will need some minor modification as we update to azure_identity 0.6.0.

These are just things to keep in mind for the future.

Comment (mattphelps-8451) [reply src/auth/azure.rs:44]: @KovacicAi I think this is a solid reminder to not let perfect be the enemy of the good! This PR cleans up a lot of the code. We don't need to tackle it all at once. Let's keep these auth code flow ideas in the back of our minds for the next enhancement.

File: Cargo.toml (hunk 2 above)
Comment (mattphelps-8451) [reply Cargo.toml:36]: Ah @KovacicAi I see those oauth2 changes [here](https://github.com/Azure/azure-sdk-for-rust/pull/9980) and [here](https://github.com/Azure/azure-sdk-for-rust/pull/997). These will be good things to know when transitioning to 0.6.0.

**Comment by mattphelps-8451** [comment]
@KovacicAi Feel free to approve and merge if you want! 🚀 Do you want this to be a new release of the CLI? Fortunately doesn't look like any breaking changes. Currently we're on `0.1.0`

**Review by aidankovacic-8451 — APPROVED** [review]
